### DISCUSSION TITLE
Trying to understand the workflow

### DISCUSSION THREAD

**Discussion by lucas-mior** [OpenModelica/OpenModelica#12377] [General]
I am having a hard time grasping what is the expected way to use openmodelica.
I like that the model has a visual and textual representation, however as soon as you start doing work in the diagram,
the text becomes unmanageable because of the annotations. So I guess I should stick to using only one of the forms?
But then again, there are tools in the docs that expect that you include annotations in your model in order to use some features.

Also related, suppose I have a block that I implement myself. Am I really supposed to position the visual connectors for each real input/output that I have? And then, if I want to know what variable I am connecting, I need to remember each connector position or
add visual artifacts to make it clear? At that point it seems easier to just wire everything by hand (writing `x = y;` for each connection) instead of relying on the diagram interface.

As an example of a better interface, GTSuite allows you to connect any block with any other block, and you select which specific variables are connecting afterwards, so you know exactly, without relying on connector positions (you also dont have do draw/position visual connectors).

**Comment by casella** [comment]
Regarding your first question, the Modelica specification does not require to only use graphical vs. textual modelling. However, it is normally good practice to avoid mixing them up, otherwise it becomes difficult to understand the code. My guidelines are the following:
-  basic models are written by declaring variables and writing variables and equations or initial equations in the text layer- you only use the GUI to draw the model icon and to drag and drop top-level connectors on the icon layer, so they get their graphical annotation automatically
- sub-system and system models are then built by dragging, dropping and connecting components, so the equation section only contains connect statements
- if you need some loose equations to compute some extra quantities in a graphically built model, you can use the Modelica.Blocks.Components.Source.RealExpression model: you can type in any expression and then you get the result on an output connector that can be used as input for other blocks

**Comment by lucas-mior** [comment]
RealExpression really does help, thank you. I wish I could explictly connect the inputs, though.
I guess it is the way it is.

**Reply by casella** [reply]
I'm not sure I understand exactly what you mean by "connecting the inputs". Can you make a concrete example?

BTW, I am not familiar with GTSuite, but is would be relatively straightforward to program a GUI that works like that with OpenModelica.

**Reply by lucas-mior** [reply]
I meant that using RealExpression, any variable could be used without explict knowledge. One would have to look into the equation to make sure a value is not being used. Not a big deal.

**Reply by casella** [reply]
Yes, but you could easily get into spaghetti modelling. Good luck debugging that, if you end up with one equation too many. Using structured models with balanced connectors allows to manage complex models much more reliably.

**Reply by lucas-mior** [reply]
Yes, that's the current behaviour. RealExpression allows accessing any variable. I wish there was some way to guard against that, without having to write a whole new model, draw connectors, etc, for a single expression.

**Reply by casella** [reply]
If you want to prevent using RealExpression to access any variable, you can use protected variables in your component models and be very strict in what is actually accessible from the outside.

On the other hand, Real expression provide an output connector only, that needs to be explicitly connected to one (or more) input connectors. So, the Real expression can "read" anything which is not protected, but it cannot "write" inside models, that can only be done through the proper input connectors.

**Comment by casella** [comment]
BTW, using connections instead of plain equations like x = y has the advantage that you can develop balanced components, which are guaranteed to form balanced systems when connecting them together. See [MLS Section 4.7](https://specification.modelica.org/maint/3.6/class-predefined-types-and-declarations.html#balanced-models). This helps avoiding the unfortunate situation when you end up with a system having 12345 equations and 12346 variables, and you have no clue where the missing equation is. With balanced components, you can check them one by one, bottom-up or top-down (though the top-down diagnostic check is currently not yet implemented in OpenModelica)
